Add draft API for querying models

bmerry opened the pull request:
There is no implementation; this is just to gather feedback on the API.

bmerry commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+    UNPOLARIZED_POWER = 4
+
+
+class Parameter:
```
This class is future-proofing for things like elevation/temperature-dependent models. I don't expect it to see any use to start with.

bmerry replied:
I'm also thinking of just deleting it for now; it'll be easy enough to add back in once we start shipping parametrized models.

bmerry replied:
Deleting it now ensures it won't be used. As is, the API allows unused parameter values to be passed.

bmerry commented:
I've set the target branch to `primary-beam`, with the idea that primary beam work will be merged there over a number of PRs and only merge into `master` once it's usable.

bmerry commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+        `m`) with a frequency. The dimensions of the output will be first those
+        of `frequency`, then those of `l` and `m` (which are broadcast with
+        each other), and finally the row and column for matrices if
+        `output_type` is one of the matrix types.
```
Feedback on the axis order is welcome. katholog works with the matrix indices as the leading (slowest-varying) axes, but my feeling is that for computation (as opposed to visualisation) one will need to work with the whole matrix rather than individual terms.

bmerry commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+        parameters
+            Additional parameters to the model. Parameters that are not used
+            are silently ignored.
+
```
I'm going to add an optional `out` parameter that lets the user provide the memory. It looks like memory allocation/faulting is a significant part of the execution time so recycling memory should help.

ludwigschwardt reviewed (commented):
How would lazy loading impact the API? Is it transparent?

ludwigschwardt commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+    def from_sky_coord(self, target: SkyCoord) -> 'RADecFrame':
+        """Generate a frame from a target (assuming an AltAz mount).
+
+        The `target` must have ``obstime`` and ``obslocation`` properties, and
```
Astropy calls it `obstime` and `location`.

bmerry commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+        """Parameters of the model."""
+        raise NotImplementedError()
+
+    def sample(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noqa: E741
```
In my head I had the output axis order as frequency, m, l (so that the horizontal coordinate is the fastest-varying, as is conventional for images), although I see the documentation doesn't actually say that.

ludwigschwardt replied:
Ah right, then I would leave the input parameters as is, because `(m, l)` is going to throw some folks on the input side, and then the input is at least the reverse of the output.

bmerry commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+
+    model_type: ClassVar[Literal['primary_beam']] = 'primary_beam'
+
+    def resolution(self, frequency: u.Quantity) -> float:
```
I had also thought of doing that and didn't have an opinion either way, so since you thought it too I'll do it.

bmerry commented:
> I think the FITS beams are separate images per pol matrix element.

Tom, what has been your experience?

bmerry commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+        raise NotImplementedError()
+
+    def min_radius(self, frequency: u.Quantity) -> float:
+        """Maximum distance from the pointing centre at which model has full coverage."""
```
Yes, I did have trouble with the names. Basically if you have a square model, there is distance from the centre to an edge, and distance from the centre to a corner. I'm not 100% sure if both are going to be useful.

ludwigschwardt replied:
I guess the mathematically appropriate terms would be `inradius` and `circumradius`. At least you are forcing the user to look it up instead of guessing what it means :joy:

bmerry replied:
I like that actually.

ludwigschwardt replied:
👍🏾

bmerry commented on src/katsdpmodels/primary_beam.py:
```diff
@@ -0,0 +1,256 @@
+
+
+class RADecFrame:
+    """Coordinate system aligned with the celestial sphere.
```
It could really be any RA/Dec frame as long as you supply a corresponding parallactic angle (I'm ignoring the distortions due to aberration since I think they're probably much smaller than the details in the beam).

bmerry replied:
Maybe I should change "the celestial sphere" to "a celestial sphere", and clarify it in the details?

bmerry commented:
> How would lazy loading impact the API? Is it transparent?

It should be transparent.

bmerry commented:
@ludwigschwardt I think I've addressed most of your comments. I'm going to leave axis order until I get more feedback from others. If there is no clear winner it may be sensible to return whatever is cheapest to implement and leave the user to reorder it (which is only a few lines of code, although possibly expensive in memory bandwidth).

ludwigschwardt reviewed (approved):
Happy for now. Let's hear from the rest :-)

mauch commented:
> > I think the FITS beams are separate images per pol matrix element.
> 
> Tom, what has been your experience?

Unfortunately there isn't any standard way to represent the beam in FITS across the imagers I've been using.

Both DDFacet and MFBeam take a separate image file per Jones matrix element and per real and imaginary parts of the complex number (sadly there is no `BITPIX` in FITS for complex values). Both use the 'standard' 4 axes for radio images, with [Stokes/Pol, frequency, m, l] order (The `STOKES` axis has length 1 in this case). My vote would be to stick to this 'standard' ordering of the axes.

wsclean is more non-standard - it takes a single 6 dimensional FITS file with order [time, frequency, antenna, 'matrix', m, l] - the 'matrix' axis has 4 elements [real XX, imaginary XX, real YY, imaginary YY] (Only diagonal Jones terms are supported at present). wsclean also has its own beam models for LOFAR & MWA and perhaps a few others (from the [EveryBeam](https://git.astron.nl/RD/EveryBeam) package ) - but MeerKAT is not on the list at present as far as I'm aware.

I've never tried gridding with the beam in CASA - but my understanding is that CASA will only use its internal beam models for VLA and ALMA. MeerKAT is not supported, and as far as I know there is no way for the user to provide their own models - but I'm not a CASA expert so I'm not sure how these models are structured.

mauch reviewed (approved):
Looks good to me so far.

bmerry commented:
Since there doesn't seem to be a clear argument for one axis ordering over another, I'm going to leave it as is for now and change it if it's going to be cheaper to implement something else. Ultimately it's quite easy for the user to reorder axes, albeit potentially expensive if the memory access pattern becomes unfriendly.

@mattieudv @SihlanguI since this has been open for several weeks I assume you don't have any comments at this time and I'm going to merge. You can still add comments later but I reserve the right to say that it's too late to change because it would mean rewriting the implementation :-)
